status/controls.py

```diff
@@ -0,0 +1,110 @@
+                headers=headers,
+                ws_data = ws_data,
+                lims_uri=BASEURI,
+            )
+        )
+
+    def find_control_data(self, control_type):
+        """Find control data from the couchDB from project/controls view
+        """
+        from collections import defaultdict
+        result = defaultdict(dict)
+        controls_view = self.application.projects_db.view("project/controls")
+
+        all_cont_proj = controls_view[[control_type,'']:[control_type,"Z"]]
+        for cont_proj in all_cont_proj:
+            for cont_sample in cont_proj.value:
+                for workset in cont_proj.value[cont_sample]: # here we create one entry in result for each workset, this will be one line in the controls table
+                    if workset != "no_workset":
+                        result[workset]["sample_id"] = cont_sample
+                        result[workset]["customer_name"] = cont_proj.value[cont_sample][workset]["customer_name"]
+                        if "status_manual" in cont_proj.value[cont_sample][workset]: # status originates from LIMS project overview, is often not set for controls
+                            result[workset]["status_manual"] = cont_proj.value[cont_sample][workset]["status_manual"]
+                        else:
+                            result[workset]["status_manual"] = "* In Progress" # asterisk indicates that the status in LIMS is not set, the sample has a workset and so MUST be at least "In Progress"
+                        result[workset]["project"] = cont_proj.key[1]
+                        result[workset]["workset_name"] = cont_proj.value[cont_sample][workset]["workset_name"]
+                        if not "workset_id" in cont_proj.value[cont_sample][workset]:
+                           result[workset]["workset_id"] = "NA"
+                        else:
+                            result[workset]["workset_id"] = cont_proj.value[cont_sample][workset]["workset_id"]
+                        if "prep_status" in cont_proj.value[cont_sample][workset]:
+                            result[workset]["prep_status"] = cont_proj.value[cont_sample][workset]["prep_status"]
+                        else:                            
+                            result[workset]["prep_status"] = "" 
+                        result[workset]["sequenced_fc"] = cont_proj.value[cont_sample][workset]["sequenced_fc"]
+        return result
+    
+    def worksets_data(self):
+        """retrieves projects for each workset and return a dictionary:
+        {[workset_id, workset_name]: {'ws_project_names':[P.roject_00_01, P.roject_00_02, ...], 'ws_project_numbers':[P12346, P23456, ...]}}
+        be aware that the workset_id and workset_name are concatenated in the key and that workset_id may not be present in the workset document.
+        The project names and project numbers are in the same order.
```

Comment:
What if the order becomes scrambled for whatever reason? Wouldn't a connected datatype be more resistant to issues? Like `{ 'ws_projects': [{'P.roject_00_01': 'P12346'}, {'P.roject_00_02': 'P23456'}, ...]`

Reply:
I do not see how the order can become scrambled to be honest, and went with a structure that would be marginally easier to work with downstream. But I understand that you would want to be rather save than sorry. I can change it.